### ISSUE TITLE
opt: normalize sequential `%` in regexes to single `%`

### ISSUE THREAD

**Issue by mgartner** [cockroachdb/cockroach#80192]
There are several optimization the execution engine can perform when evaluating `LIKE` patterns with `%`, which matches zero or any number of any character. For example, see #80073.

Sequential `%`s can prevent these optimizations, but they are semantically equivalent to a single `%`, so the optimizer should normalize them to a single `%`.

For example:

```sql
-- The following expressions are equivalent:
col LIKE '%%abc%%%%'
col LIKE '%abc%'
```

Jira issue: CRDB-15814

**Comment by ADustyOldMuffin** [comment]
I'd be interested in taking this on.

**Comment by GV14982** [comment]
Is this issue still up for grabs?

**Comment by ADustyOldMuffin** [comment]
You can take a look at my PR and if you can solve it from the feedback there then sure! I haven't had time to learn the custom language to find the area to place the fix.

**Comment by mgartner** [comment]
Theses pointers may be helpful: https://github.com/cockroachdb/cockroach/pull/80432#issuecomment-1109849896

**Comment by rishabhkumar296** [comment]
Hi, I am actively working on this issues.

**Comment by dils2k** [comment]
@mgartner I am curious if the solution should also handle scenarios involving `LIKE ... ESCAPE ...` comparisons. I am asking because, as far as I understood, LIKE comparison with ESCAPE has a different internal representation than a simple LIKE comparison, and would likely have a different implementation of this normalization.

```sql
root@localhost:26257/defaultdb> explain (opt) select '%' || (random()::text) like '!%';
                     info
-----------------------------------------------
  values
   └── (('%' || random()::STRING) LIKE '!%',)
(2 rows)

Time: 1ms total (execution 1ms / network 0ms)
```
```sql
root@localhost:26257/defaultdb> explain (opt) select '%' || (random()::text) like '!%' escape '!';
                           info
-----------------------------------------------------------
  values
   └── (like_escape('%' || random()::STRING, '!%', '!'),)
(2 rows)

Time: 1ms total (execution 1ms / network 0ms)
```

**Comment by mgartner** [comment]
@dils2k Correct, we currently convert `LIKE ... ESCAPE ...` into the `like_escape` built-in function. This PR doesn't currently cover that case. The primary reason for normalizing sequential wildcards is to unlock other optimizations which they can prevent, like #80073 and index-acceleration. Those types of optimizations do not currently support `LIKE/ESCAPE` as-is, so supporting `LIKE ... ESCAPE ...` here wouldn't provide benefit.

**Comment by dils2k** [comment]
@mgartner I would like to work on this issue. I opened a PR #158025 and I would appreciate if someone could take a look and tell if I am going in the right direction.
